### ISSUE TITLE
Is 2.4.0 binary compatible with 2.3.x stream?

### ISSUE THREAD

**Issue by dotta** [brettwooldridge/HikariCP#360]
(I've failed to find a mailing list where I could ask the question. Please, feel free to close this ticket and redirect me to it if one exists.)

Hi,

I'm wondering if the 2.4.0 release is binary compatible with 2.3.x. And, more in general, if you follow [semantic versioning](http://semver.org/).

Thanks!

**Comment by dotta** [comment]
Ah, actually, it looks like I already have an answer, as I'm seeing the following problem after upgrading to 2.4.0

    java.lang.NoSuchMethodError: com.zaxxer.hikari.HikariConfig.setReadOnly(Z)V

Therefore, 2.4.0 is definitely not binary compatible with 2.3.x.

Still, I'm interested to know if patch releases are guaranteed (or not) to be binary compatible.

**Comment by dotta** [comment]
My bad. I just noticed that the project where I tried to bump the HikariCP version uses a module that depends on `HikariCP-java6`. Hence, the linkage error I reported above is not surprising.

Hence, my initial question about semantic versioning still stands. Thanks again, and sorry for the noise.

**Comment by johdah** [comment]
I got the same error when I tried to update to HikariCP 2.4 in a Play 2.4 project.

**Comment by dotta** [comment]
@johdah What db module where you using? play-jdbc? play-slick?

If it was play-slick, then that's expected, since the exception in my [above comment](https://github.com/brettwooldridge/HikariCP/issues/360#issuecomment-126607673) happened after bumping the version of hikaricp in play-slick.

**Comment by johdah** [comment]
@dotta I use play-jdbc.

**Comment by dotta** [comment]
@johdah Thanks for the info. Then, since play-jdbc uses the hikaricp jar for java8, it seems that 2.4.0 is not binary compatible with the 2.3.x stream.

**Comment by johdah** [comment]
@dotta To clarify for myself, you are talking about HikariCP-2.3 and 2.4 right? It's a little bit hard to follow sometimes since the same version numbers is used when talking about Play :confused: 

I don't really need Hikari-2.4 since I think my problems are in Play so I guess I am fine with being stuck on 2.3.x until the Play team upgrades.

**Comment by dotta** [comment]
@johdah Yes, I meant HikariCP-2.4.0 seems not to be binary compatible with HikariCP-2.3.x stream. Sorry for the confusion.

Also, I'm really sorry for the noise I'm generating in this ticket. Discussing this in a mailing list would be much better. Hopefully the HikariCP team will forgive me.

**Comment by brettwooldridge** [comment]
@dotta @johdah HikariCP 2.4.0 should be binary compatible with 2.3.x, unless deprecated APIs were being used.  API's that were removed between 2.3.x and 2.4.0 are:
```java
HikariConfig.get/setConnectionCustomizerClassName()
HikariConfig.get/setConnectionCustomizer()
```
Other than those removed methods, there should be no API incompatabilties.

Of course, the other change between 2.3.x and 2.4.0 that you noted is that Java 6 is no longer supported in the 2.4.0 release stream.

We do not currently follow semantic versioning, because the API has been too unstable and we would be on version 50.11.0 by now.  We generally follow the pattern of leaving @deprecated methods in the API for 6 months or so (however many releases that is).  We're getting to the point where it might make sense to start semantic versioning, but I'm not willing to commit to it in writing.

**Comment by szeiger** [comment]
The standard for Scala nowadays is to keep everything with the same epoch + major binary compatible. That's what we use for Play and Slick, for example. A similar scheme (enforced by running a tool like MiMa as part of the CI builds) would be useful for HikariCP so that users can rely on these semantics but you;re still free to change the API relatively frequently.

BTW, what constitutes the API? I assume it's HikarciConfig and HikariDataSource. Anything else?

**Comment by dotta** [comment]
@brettwooldridge First, thanks for the link to the mailing list, and sorry for missing it. Let me know if you want me to move the discussion there.

I just run the [japi-compliance-checker](https://github.com/lvc/japi-compliance-checker/) (a tool for checking both binary and source incompatibilities across two version of the same library) to compare HikariCP-java6 v2.3.8 and v2.3.9, and it reported a few binary incompatibilities ([here is a handy link](http://s000.tinyupload.com/download.php?file_id=79251882567654631359&t=7925188256765463135929819) to the report) - it says the two jars are ~37% incompatible.

Based on what you are saying, until semantic versioning (or any alternative versioning policy) is adopted, there are no binary compatibility guarantees across releases.

**Comment by brettwooldridge** [comment]
@szeiger current public API is ``HikariConfig``, ``HikariDataSource``, ``HikariConfigMXBean`` and ``HikariPoolMXBean``.  And at some point ``MetricsTrackerFactory`` and ``MetricsTracker``, once they've stabilized and we publicly document their use.

**Comment by brettwooldridge** [comment]
@dotta Thanks for the report, I'll looking to that tool.  It does need to be restricted to the "public" (e.g. documented API) of the clases above.
